### PR TITLE
Fixing major bugs

### PR DESCRIPTION
**Pull request by vhavlena**
This PR fixes a couple of bugs involving:

- relevancy propagation
- storing axiomatized terms
- theory rewriter (suppress some optimizations that are not suitable for us)
- length fixing

### CODE REVIEW

File: src/ast/rewriter/seq_rewriter.cpp [hunk 1]
```diff
@@ -4527,16 +4527,16 @@ br_status seq_rewriter::mk_str_in_regexp(expr* a, expr* b, expr_ref& result) {
...
+    // if (get_re_head_tail(b, hd, tl)) {
+    //     SASSERT(re().min_length(hd) == re().max_length(hd));
+    //     expr_ref len_hd(m_autil.mk_int(re().min_length(hd)), m()); 
+    //     expr_ref len_a(str().mk_length(a), m());
+    //     expr_ref len_tl(m_autil.mk_sub(len_a, len_hd), m());
+    //     result = m().mk_and(m_autil.mk_ge(len_a, len_hd),
+    //                         re().mk_in_re(str().mk_substr(a, zero(), len_hd), hd),
+    //                         re().mk_in_re(str().mk_substr(a, len_hd, len_tl), tl));
+    //     return BR_REWRITE_FULL;
+    // }
```
Comment (jurajsic) [inline src/ast/rewriter/seq_rewriter.cpp:4539]: why is this commented out?

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 2]
```diff
@@ -335,11 +335,12 @@ namespace smt::noodler {
             expr_ref len(m_util_a.mk_numeral(rational(l), true), m);
 
             literal lit(mk_eq(len_str, len, false));
-            ctx.mark_as_relevant(lit);
+            // ctx.mark_as_relevant(lit);
```
Comment (Adda0) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:338]: Just to check again, is this part of the relevancy problems fix and is this change intended?

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 3]
```diff
@@ -335,11 +335,12 @@ namespace smt::noodler {
...
             ctx.mk_th_axiom(get_id(), 1, &lit);
         } else {
             // build axiom 1: Length(a_str) >= 0
-            {
+            { 
+                return; // TODO: fix some day
```
Comment (Adda0) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:343]: Maybe write a quick note what is to be fixed here.

**Review by Adda0 — APPROVED** [review]
Other than what Juraj mentioned, the changes seem good to me.

File: src/ast/rewriter/seq_rewriter.cpp (hunk 1 above)
Comment (vhavlena) [reply src/ast/rewriter/seq_rewriter.cpp:4539]: This seems to be the code that performs replacing of certain regexes to `substr` predicate. This is not beneficial for us because we aim at the regular constraints.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 2 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:338]: I think so. The main problem was somewhere else however.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 3 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:343]: ok

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 2 above)
Comment (Adda0) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:338]: That is good enough for me.

**Comment by vhavlena** [comment]
Should I merge it now?

File: src/ast/rewriter/seq_rewriter.cpp (hunk 1 above)
Comment (jurajsic) [reply src/ast/rewriter/seq_rewriter.cpp:4539]: Maybe just change the first line to `if (false && get_re_head_tail(b, hd, tl)) { \\FIXME-NOODLER: replacing regexes to substr is not beneficial for noodler` so we change as little lines of z3 code. It would help when we will want to update the z3 version for noodler.
Or maybe, is it possible to check here if str solver is set to noodler? So the condition would be ``if (!str_solver_is_noodler && get_re_head_tail(b, hd, tl))`.

Comment (Adda0) [reply src/ast/rewriter/seq_rewriter.cpp:4539]: We should probably add something like `FIXME: noodler substring temp fix` so that we can find this place later.

**Comment by Adda0** [comment]
I would add the mentioned FIMXE. Otherwise, I am good with merging the PR.

File: src/ast/rewriter/seq_rewriter.cpp (hunk 1 above)
Comment (vhavlena) [reply src/ast/rewriter/seq_rewriter.cpp:4539]: you are right

Comment (vhavlena) [reply src/ast/rewriter/seq_rewriter.cpp:4539]: I am not aware of the second option. I guess it could work, but it would require some effort

**Review by jurajsic — APPROVED** [review]
I think you can merge
